### ISSUE TITLE
Autosummary: ``autosummary_ignore_module_all`` does not work as documented

### ISSUE THREAD

**Issue by ClementPinard** [sphinx-doc/sphinx#10809]
### Describe the bug

As per documentation :

> Note that if an imported member is listed in `__all__`, it will be documented regardless of the value of autosummary_imported_members.

So the way I understand it is that if I have a small package like this :

```
my_package/
    __init__.py
    my_class.py
    my_module/
        my_submodule.py
```

with the following files :

`my_class.py`
```python
class myClass:
    """" my Class docstring """"
    def my_class_method(self, a):
        """ method docstring """
        return
```

`my_submodule.py`
```python
def f():
    """ f docstring """
    return
```

`__init__.py`
```python
from .my_class import myClass
from . import my_module

__all__ = ["myClass", "my_module"]
```

and a rst file like this:

```rst
An example of recursive module documentation
====================================

.. autosummary::
    :toctree: generated
    :recursive:

    my_package
```

assume the option `autosummary_ignore_module_all` is set to `False` in `conf.py`, the autosummary should only consider `myClass` and `my_module` to be part of the package `my_package`. The `generated` folder should normally look like this:

```
generated
    my_package.myClass
    my_packge.my_module
```

But actually, it looks like this :

```
generated
    my_package.my_class
```

Meaning it ignored the imported elements, even though they were in the `__all__` and it did take the `my_module` package even though it was not in the `__all__`

### How to Reproduce

```
$ git clone git@github.com:ClementPinard/sphinx_autosummary_bug_example.git
$ cd sphinx_autosummary_bug_example.gt
$ pip install sphinx
$ sphinx-build -b html docs docs/_build
$ # see docs/generated/ folder
```

### Expected behavior

generated folder should have the files

```
my_package.rst
my_package.my_module.rst
my_patckage.my_module.my_submodule.rst
my_package.myClass.rst
```

### Your project

https://github.com/ClementPinard/sphinx_autosummary_bug_example

### OS

Ubuntu 20.04

### Python version

3.9

### Sphinx version

5.1

### Sphinx extensions

sphinx.ext.autodoc, sphinx.ext.autosummary

**Comment by ClementPinard** [comment]
After some investigation in the code, I do believe that there is a problem in the generate.py file, when populating the `ns` dictionnary

https://github.com/sphinx-doc/sphinx/blob/5.x/sphinx/ext/autosummary/generate.py#L313


## Problem with `myClass`

`ns["members"]` has the right dict, i.e. `myClass` and ̀my_module`, but when trying to populate `ns["classes"]` , `myClass` gets dismissed, because `imported_member` is not  ̀True`, and the function `get_members` does not check if the imported memeber is in `__all__` or not as seen here : https://github.com/sphinx-doc/sphinx/blob/7473b05e0d7142ccd6196b2fd61a4f6c9dc21fbc/sphinx/ext/autosummary/generate.py#L259

See how if skips the block if either the `imported` is not evaluated to `True` or the `__module__` of considered object is not the root, i.e. the element is imported.

As a consequence, `ns["classes"]` is empty

 ## Problem with `my_module`

The code only seem to care about modules found by the function `https://github.com/sphinx-doc/sphinx/blob/7473b05e0d7142ccd6196b2fd61a4f6c9dc21fbc/sphinx/ext/autosummary/generate.py#L320` and otherwise are simply ignored.

It explains why we find `my_package.my_class` even though it's not in `my_package.__all__`. Not sure how to explain the lack of `my_package.my_module` though, it seems that the fact that it has a submodule and no `__init__.py` is the problem.

**Comment by ClementPinard** [comment]
See a proposition for how to fix it here : #10811
